doc/compiler/osr/OSR.md

```diff
@@ -0,0 +1,237 @@
+                                                  ---------------------    ---------------------
+After OSRGuardinsertion:
+
+                    --------------------------
+                    |     entry              |---> catch 1
+                    --------------------------
+                             |
+                    --------------------------
+                    |   OSR guard            |
+                    --------------------------
+                   /         |
+                  /          |
+  ----------------  --------------------------
+  |jitInduceOSR  |  |  non-overriden guard 1 |---> catch 1
+  |AtCurrentPC   |  |                        |
+  ----------------  --------------------------
+        |          /         |
+    catch 1       /          |
+                 /           |
+  ---------------   --------------------------
+  |call B(cold) |   |    call C (from B)     |---> catch 2
+  ---------------   --------------------------
+       |         \           |
+    catch 1       \          |
+                    --------------------------
+                    |                        |---> catch 1
+                    --------------------------
+                             |
+                    --------------------------    ---------------------    ---------------------
+                    |     exit               |    | catch 2           |    | catch 1           |
+                    --------------------------    ---------------------    ---------------------
+                                                          |                         |
+                                                  ---------------------    ---------------------
+                                                  |prepareForOSRHelper|___>|prepareForOSRHElper|___>  exit
+                                                  |   iconst 0        |    |  iconst -1        |
+                                                  ---------------------    ---------------------
+
+```
+
+
+## Supporting Classes
+
+**OSRGuardAnalysis**  : detects OSR guards that are no longer required
+
+**OSRGuardInsertion** :  replaces HCR guards with OSR guards based on the fear analysis described above.
+
+**OSRGuardRemoval** : removes OSR guards that became redundant due to optimizations after OSRGuardInsertion. Uses OSRGuardAnalysis
+
+
+## nextGenHCR Control Options
+
+As this feature relies on HCR and OSR, both will be enabled when in nextGenHCR.
+To prevent these from being enabled and to avoid any additional modifications to the trees, use -Xjit:disableNextGenHCR.
+In the event that disableOSR or the environment variable TR_DisableHCR is set, nextGenHCR will disable itself.
+It should be noted that if the VM requests HCR when nextGenHCR is disabled, the traditional implementation will be used.
+Setting the environment variable TR_DisableHCRGuards will result in nextGenHCR performing as usual,
```

Comment:
I'm not sure what you mean by *performing as usual*

Reply:
This is an old text :) Let me see how it can be improved.

Reply:
This is from the old doc. I think the next sentence explains: "This is a debug mode to evaluate OSR overhead and it will not correctly implement redefinitions." I guess "usual" means "as described above". I will replace.

Reply:
Oh, I understand. In that case, what about *working* as described above? In this context *performing* sounds to me like it's about the performance of the generated code